### PR TITLE
 🌱Added function to update control plane endpoint

### LINKED ISSUE #977: Add utility function to update controlPlaneEndpoint
## Description
Update the current code of the HetznerCluster controller (https://github.com/syself/cluster-api-provider-hetzner/blob/main/controllers/hetznercluster_controller.go#L202-L225)  with creating a separate utility function that contains the existing code and which gets the hetznerCluster as input

Then unit test this utility function accordingly.

### PR DESCRIPTION
**Pull request by razashahid107**
**What this PR does / why we need it**:
This one fixes a part of the issue #977. Added utiltiy function to update control plane endpoints. 
**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #977 

**Special notes for your reviewer**:
This is not the complete PR. I'll add the unit tests. I request to get my function reviewed. I want to know what I can make better with my function. I'm currently feeling a little unsure about the false condition that I've added to the end of the function which can potentially set the status of the hetzner cluster to false. Is it okay if it stays that way? Or do I edit it up? I'd love feedback. 

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._
No it does not
**TODOs**:

- [ ] squash commits
- [ ] include documentation
- [ ] add unit tests

### CODE REVIEW

File: controllers/hetznercluster_controller.go [hunk 1]
```diff
@@ -260,6 +238,33 @@ func (r *HetznerClusterReconciler) reconcileNormal(ctx context.Context, clusterS
...
 }
 
+func SetControlPlaneEndpoint(hetznerCluster *infrav1.HetznerCluster) (bool) {
+	if hetznerCluster.Spec.ControlPlaneLoadBalancer.Enabled {
+		if hetznerCluster.Status.ControlPlaneLoadBalancer.IPv4 != "<nil>" {
+			defaultHost := hetznerCluster.Status.ControlPlaneLoadBalancer.IPv4
+			defaultPort := int32(hetznerCluster.Spec.ControlPlaneLoadBalancer.Port)
+
+			if hetznerCluster.Spec.ControlPlaneEndpoint == nil {
+				hetznerCluster.Spec.ControlPlaneEndpoint = &clusterv1.APIEndpoint{
```
Comment (razashahid107) [inline controllers/hetznercluster_controller.go:248]: Just noticed with my current code, this line won't take effect, because the objects are passed by value and not by reference hence the Host and Port will never get set right? 
I was thinking maybe I can just pass the hetznerCluster parameter by reference and then use it in the whole function by reference, this way I don't need to return any value either. Is this approach okay?

Comment (janiskemper) [reply controllers/hetznercluster_controller.go:248]: you store everything in hetznerCluster which is a pointer. therefore it should take effect!

Comment (janiskemper) [reply controllers/hetznercluster_controller.go:248]: and yes, the approach is okay so far!

Comment (razashahid107) [reply controllers/hetznercluster_controller.go:248]: Okay so I do not need to alter with the SetControlPlaneEndpoint() function? 
I can begin with the unit test then right?

Comment (janiskemper) [reply controllers/hetznercluster_controller.go:248]: if you didn't change any logic, then all is well! Some minor cosmetics I would change. But go ahead with unit tests for now!

Comment (janiskemper) [reply controllers/hetznercluster_controller.go:248]: I would use the standard Go library for the unit tests of this function, instead of Gingko/gomega we use with envtest

Comment (razashahid107) [reply controllers/hetznercluster_controller.go:248]: Aight I'll get started with the unit tests. Also thank you for making time to review the PR Janis.

**Comment by razashahid107** [comment]
Hi @janiskemper . Okay basic question. I have worked with a few unit tests earlier however they were for relatively simpler Go code probably hence they contained functions such as func Testxxx(). Now going through hetznercluster_controller_tests.go its a little different. I've tried going through the test file and I'm still understanding things, but can you maybe just point me to the right direction with where do I start. 
Like do I just create a function like func TestSetControlPlaneEndpoint() and maybe proceed to write code there? I'm sorry I just feel a little lost and I feel a little direction would really help.

**Comment by janiskemper** [comment]
that's exactly why I said
> I would use the standard Go library for the unit tests of this function, instead of Gingko/gomega we use with envtest

yes you can and should proceed like  that

**Comment by razashahid107** [comment]
Oh okay okay. Aight. I'll proceed this way. Thank you Janis

**Comment by razashahid107** [comment]
Hey @janiskemper can you please go through the below code? 
So I'm basically trying to pass the hetznercluster as input to the function, now what I'm not able to understand is how do I pass it as an argument in the code below? I'm sure there is something very basic I'm missing. I'd love if you can help out. 

> tests := []struct{
		name string
		args args
		want bool
	}{
		{
			name: "should return true",
			args: args{hetznerCluster},
			want: true,
		},
		{
			name: "should return false",
			args: args{hetznerCluster},
			want: false,
		},
	}

**Comment by janiskemper** [comment]
I'm not sure what "args" is in your code...

**Comment by razashahid107** [comment]
Oh I'm sorry. I had created a type which includes the hetznercluster. Wait I'll put in the code: (cc @janiskemper)

> type args struct {
		hetznerCluster *infrav1.HetznerCluster
	}
	tests := []struct{
		name string
		args args
		want bool
	}{
		{
			name: "should return true",
			args: args{hetznerCluster},
			want: true,
		},
		{
			name: "should return false",
			args: args{hetznerCluster},
			want: false,
		},
	}

**Comment by razashahid107** [comment]
Hey @janiskemper . I'd really appreciate if you can provide feedback on the latest commit. As of now I've kept the two test cases almost same, I understand I will have to make changes to them. But I feel a little lost as to how do I make updates with the hetznercluster parameter. Like how do I set a case in which the test returns false, and how do I set a case in which a test returns true, this is what is confusing me. I hope I'm able to convey my point. 

I also hope I am going in the right direction. I'd really appreciate your feedback.

**Comment by janiskemper** [comment]
From the code side you have created a pattern that you can use to write tests. But I would say that you struggle right now what you should test.. 
Well, pretty much tests that cover most of or even all of the logic that the function contains. 

According to the pattern 
if I do xyz, then I expect abc.
If i do xyz2, then I expect abc2. 

One shortcoming of your current approach is that not only it is interesting which boolean the function returns, but also the changes to the hetznerCluster have to be validated. 

I suggest that you first start with a list of tests that you wanna implement... Not writing code yet, just what would make sense to do.

**Comment by razashahid107** [comment]
```
func SetControlPlaneEndpoint(hetznerCluster *infrav1.HetznerCluster) (bool) {
	if hetznerCluster.Spec.ControlPlaneLoadBalancer.Enabled {
		if hetznerCluster.Status.ControlPlaneLoadBalancer.IPv4 != "<nil>" {
			defaultHost := hetznerCluster.Status.ControlPlaneLoadBalancer.IPv4
			defaultPort := int32(hetznerCluster.Spec.ControlPlaneLoadBalancer.Port)

			if hetznerCluster.Spec.ControlPlaneEndpoint == nil {
				hetznerCluster.Spec.ControlPlaneEndpoint = &clusterv1.APIEndpoint{
					Host: defaultHost,
					Port: defaultPort,
				}
			} else {
				if hetznerCluster.Spec.ControlPlaneEndpoint.Host == "" {
					hetznerCluster.Spec.ControlPlaneEndpoint.Host = defaultHost
				}
				if hetznerCluster.Spec.ControlPlaneEndpoint.Port == 0 {
					hetznerCluster.Spec.ControlPlaneEndpoint.Port = defaultPort
				}
			}
			return true
		}
	} else if hetznerCluster.Spec.ControlPlaneEndpoint != nil {
		return true
	}
	return false
}
```
Okay so according to the code above I feel we'd want a few specefic test cases to take place which return true:

1. So if the Loadbalancer is enabled and its IPV4 holds a string value of "nil" then the test should by default just return true (a few things will get set in the if condition, but they do not matter to us, what matters is the return value which will be true). 
2. Otherwise if that is not true and if the ControlPlaneEndpoint is nil then we should also return true (just that this time we did not set the values for other things as we did in the previous if statement)

If none of these match then the function should return false. 

Seeing the above conditions, I think I should include three test cases. However, the problem still persists, how should I mention these conditions when writing the unit test? I dont think its supposed to be mentioned in the tests struct or is it?

**Comment by janiskemper** [comment]
I believe that there are a lot more possible tests. Pretty much every if-else statement gives you another possibility to test something

**Comment by razashahid107** [comment]
Oh, I'll look into this. I thought I'd maybe base my test cases solely on the return statements. 
Can you maybe suggest an example for the other test cases possible? I can maybe take direction from that

**Comment by razashahid107** [comment]
@janiskemper whats confusing me is that if I take the if statement 
`if hetznerCluster.Spec.ControlPlaneEndpoint == nil` and add it to the test case, what do I put in the want? Do I just put in true (I think if the function is at this stage of the program, then it will eventually return true)? Or am I mistaking?

**Comment by razashahid107** [comment]
Hi @janiskemper . So I started writing my test cases again, and I'm finding myself comfortable addressing possible cases, thank you for the activity earlier. 
I have a few questions as I write the tests. 
```
func SetControlPlaneEndpoint(hetznerCluster *infrav1.HetznerCluster) (bool) {
	if hetznerCluster.Spec.ControlPlaneLoadBalancer.Enabled {
		if hetznerCluster.Status.ControlPlaneLoadBalancer.IPv4 != "<nil>" {
			defaultHost := hetznerCluster.Status.ControlPlaneLoadBalancer.IPv4
			defaultPort := int32(hetznerCluster.Spec.ControlPlaneLoadBalancer.Port)

			if hetznerCluster.Spec.ControlPlaneEndpoint == nil {
				hetznerCluster.Spec.ControlPlaneEndpoint = &clusterv1.APIEndpoint{
					Host: defaultHost,
					Port: defaultPort,
				}
			} else {
				if hetznerCluster.Spec.ControlPlaneEndpoint.Host == "" {
					hetznerCluster.Spec.ControlPlaneEndpoint.Host = defaultHost
				}
				if hetznerCluster.Spec.ControlPlaneEndpoint.Port == 0 {
					hetznerCluster.Spec.ControlPlaneEndpoint.Port = defaultPort
				}
			}
			return true
		}
	} else if hetznerCluster.Spec.ControlPlaneEndpoint != nil {
		return true
	}
	return false
}

```
So with the if conditions above in hetznercluster_controller.go at
```
else {
  if hetznerCluster.Spec.ControlPlaneEndpoint.Host == "" {
```
what if the values are as such: Host == "xzy" and port == "1". The function will still return true. I think it should return false, because we did not set the values of Host and Port. Do we need to edit this in the function?

**Comment by razashahid107** [comment]
@janiskemper Can you please review the latest commit?

**Comment by janiskemper** [comment]
the main part of the test should also be to check the changes that have been made to the hetznercluster... that you don't test currently"

**Comment by razashahid107** [comment]
Ahan. @janiskemper can you please elaborate on that a little. Like are we talking about the changes to `hetznerCluster.Spec.ControlPlaneEndpoint.Host` and `hetznerCluster.Spec.ControlPlaneEndpoint.Port`?

**Comment by janiskemper** [comment]
about hetznerCluster.Spec.ControlPlaneEndpoint in general, yeys

**Comment by razashahid107** [comment]
I'm sorry @janiskemper I feel a little lost as to what am I missing. I'll add another commit maybe rephrasing the test case names. I've tried addressing changes being made to the ControlPlaneEndpoint.Host and ControlPlaneEndpoint.Port in the first two test cases. Can you please go through it? 
I'd love feedback at what I'm missing

**Comment by janiskemper** [comment]
you just have to also include test cases for changes of hetznerCluster.Spec.ControlPlaneEndpoint.. At the moment you only test the return boolean in https://github.com/syself/cluster-api-provider-hetzner/pull/1004/files#diff-17dd1bcc2376d005c272189be910b2625e9a01d4bc08d71663eb972ed88f0df6R877.
Just not focus on coding for now.. Look at the function from a high level perspective and think about the logic it contains and for stuff that you might want to test for

**Comment by janiskemper** [comment]
yes, exactly!

**Comment by razashahid107** [comment]
@janiskemper. So I was earlier adding test cases, but I think the test cases are fine, like I do not need to add more. The condition I added should be okay right? Or would you recommend adding another test case?
If everything is okay, then do I just squash the commits and push?

**Comment by janiskemper** [comment]
can you maybe take a step back and think about every test that you could potentially do? Because I think this structure of the fixed table-driven test holds you back. Just make a list of tests that could be implemented. I believe we are still missing out many

**Comment by razashahid107** [comment]
- I can test if ControlPlaneLoadbalancer.IPV4 is `"<nil>"` then the variables `defaultHost` and `defaultPort` should not be nil. (I dont think I can do this though, because these are local variables and they are not being passed in the function. Or can I?)
- I can also try another test which checks if Host is empty like `hetznerCluster.Spec.ControlPlaneEndpoint.Host == ""` then as the test ends it should be equal to `hetznerCluster.Status.ControlPlaneLoadBalancer.IPv4`. 
- I can do the same thing with `hetznerCluster.Spec.ControlPlaneEndpoint.Port == 0` and check if it equals `hetznerCluster.Spec.ControlPlaneLoadBalancer.Port`
Am I approaching this correct?

**Comment by janiskemper** [comment]
just for an example: What if controlPlaneLoadBalancer is not enabled? Then other things happen.. You can really go through the whole logic and pretty much any possible combination of things that might happen can be (theoretically) tested.

**Comment by janiskemper** [comment]
you can check what happens if load balancer is not enabled and if the controlPlaneEndpoint is nil / not nil. 
You can check what happens if the load balancer is enabledand the IPv4 is "<nil>", or is not.
You can check what happens if ControlPlaneEndpoint is nil or not nil and what happens if host and port are set or not. 
You can check what happens for different values of ControlPlaneEndpoint and ControlPlaneLoadBalancer. 
You can check all of that!

**Comment by razashahid107** [comment]
🤦 . I'm sorry @janiskemper. Earlier you mentioned about checking changes made to hetznercluster, I think I got very zoomed in on that only. I was trying to avoid want statements, which is why it did not make sense to me. Can you please go through the latest commit. I feel that there can be more test cases with more values, but they'll more or less be same. Is there still something that I'm missing.

**Comment by janiskemper** [comment]
maybe you can do the following: I think the table-driven test really makes you think very one-directional. Can you completely get rid of it and do one test case after another? I think that would really help you... I mentioned a lot of tests in my previous comment that you can potentially do.

**Comment by razashahid107** [comment]
I thought I made test cases for all the test cases you suggested? I'm defo missing some major point. I can discard all my tests and restart from line one. Would you suggest I do that?

**Comment by janiskemper** [comment]
you implented table driven tests. Can you re-implement what you currently have (and maybe new tests) in a manner that doesn't use table-driven tests?

**Comment by razashahid107** [comment]
Writing a function for each test counts? Something similar to whats mentioned in this link [https://golang.cafe/blog/golang-table-test-example]

**Comment by janiskemper** [comment]
yes, but maybe a pattern that uses `t.Run("InvalidMajorVersion", func(t *testing.T) {// test code}` within a test is better.

**Comment by janiskemper** [comment]
closing as this is replaced by #1023
